semgrep-core/toy-matcher/Matcher.ml

```diff
@@ -0,0 +1,562 @@
+     if applicable
+   - add the named atom that was just matched, if applicable
+*)
+let extend env opt_ellipsis opt_name captured_sequence =
+  let env =
+    match opt_ellipsis with
+    | None -> env       (* we were not in an ellipsis *)
+    | Some None -> env  (* we were in an anonymous ellipsis *)
+    | Some (Some (ellipsis_name, acc)) ->
+        Env.add ellipsis_name (List.rev acc) env
+  in
+  match opt_name with
+  | None -> env
+  | Some name ->
+      Env.add name captured_sequence env
+
+(*
+   Turn a string into a list of chars.
+*)
+let parse s =
+  let acc = ref [] in
+  for i = String.length s - 1 downto 0 do
+    acc := s.[i] :: !acc
+  done;
+  !acc
+
+let unparse l =
+  let buf = Buffer.create (List.length l) in
+  List.iter (Buffer.add_char buf) l;
+  Buffer.contents buf
+
+let print_result oc opt_bindings =
+  match opt_bindings with
+  | None ->
+      fprintf oc "not a match\n"
+  | Some l ->
+      fprintf oc "match: {\n";
+      List.iter (fun (name, subseq) ->
+        fprintf oc "  %s: %S\n"
+          name subseq
+      ) l;
+      fprintf oc "}\n"
+
+(* to be appended to existing line *)
+let print_ellipsis oc ellipsis =
+  match ellipsis with
+  | None -> ()
+  | Some None -> fprintf oc " in-ellipsis"
+  | Some (Some (name, acc)) ->
+      fprintf oc " in-ellipsis:%s:%S" name (unparse (List.rev acc))
+
+(* to be appended to existing line *)
+let print_env oc env =
+  fprintf oc " {";
+  let is_first = ref true in
+  Env.bindings env
+  |> List.iter (fun (name, subseq) ->
+    if !is_first then
+      is_first := false
+    else
+      fprintf oc " ";
+    fprintf oc "%s:%S"
+      name
+      (unparse subseq)
+  );
+  fprintf oc "}"
+
+(* to be appended to existing line *)
+let print_pat_head oc pat =
+  match pat with
+  | [] ->
+      fprintf oc " _:''"
+  | (atom, opt_name) :: _ ->
+      (match opt_name with
+       | None -> fprintf oc " _:"
+       | Some name -> fprintf oc " %s:" name
+      );
+      match atom with
+      | Any_symbol -> fprintf oc "_"
+      | Symbol c -> fprintf oc "%C" c
+      | Ellipsis -> fprintf oc "..."
+      | Backref name -> fprintf oc "%s" name
+
+(* to be appended to existing line *)
+let print_input_head oc input =
+  match input with
+  | [] ->
+      fprintf oc " ''"
+  | symbol :: _ ->
+      fprintf oc " %C" symbol
+
+let max_trace_lines = 100
+
+(* print single line *)
+let trace_match_call ~trace stat ellipsis env pat input =
+  let match_calls = stat.match_calls + 1 in
+  stat.match_calls <- match_calls;
+  if trace then
+    if match_calls <= max_trace_lines then
+      printf "match%a%a%a%a\n"
+        print_ellipsis ellipsis
+        print_env env
+        print_pat_head pat
+        print_input_head input
+    else if match_calls = max_trace_lines + 1 then
+      printf "[exceeded max trace lines = %i]\n%!" max_trace_lines
+
+module Cache_key = struct
+  type t = ellipsis * env * pattern * input
```

Comment:
Can you add a comment explaining the intuition behind this caching idea. What do you try to cache
and why it avoids work.

Reply:
I added  something.